framework/aster-frame/src/trap/handler.rs

```diff
@@ -182,3 +183,62 @@ cpu_local! {
 pub fn in_interrupt_context() -> bool {
     IN_INTERRUPT_CONTEXT.load(Ordering::Acquire)
 }
+
+fn handle_kernel_page_fault(f: &TrapFrame) {
+    // We only create mapping: `vaddr = paddr + PHYS_OFFSET` in kernel page fault handler.
+    let page_fault_vaddr = x86_64::registers::control::Cr2::read().as_u64();
+    debug_assert!(is_kernel_vaddr(page_fault_vaddr as usize));
+
+    // Check kernel region
+    // FIXME: The modification to the offset mapping of the kernel code and data should not permitted.
+    debug_assert!({
+        let kernel_region = MemoryRegion::kernel();
+        let start = kernel_region.base();
+        let end = start + kernel_region.base() + kernel_region.len();
+        !((start..end).contains(&(page_fault_vaddr as usize)))
+    });
+
+    // Check error code and construct flags
+    // FIXME: The `PageFaultErrorCode` may not be the same on other platforms such as RISC-V.
+    let error_code = PageFaultErrorCode::from_bits_truncate(f.error_code);
+    debug!(
+        "Handling kernel page fault. Page fault address:{:x?}; Error code:{:?}",
+        page_fault_vaddr, error_code
+    );
+    debug_assert!(!error_code.contains(PageFaultErrorCode::USER));
+    debug_assert!(!error_code.contains(PageFaultErrorCode::INSTRUCTION));
+    let mut flags = PageTableFlags::empty()
+        .set_present(true)
+        .set_executable(false);
+    #[cfg(feature = "intel_tdx")]
+    {
+        // FIXME: Adding shared bit directly will have security issues.
+        flags = flags | PageTableFlags::SHARED;
+    }
+    if error_code.contains(PageFaultErrorCode::WRITE) {
+        flags = flags.set_writable(true);
+    }
+
+    // Handle page fault
+    let mut page_table: PageTable<PageTableEntry, crate::vm::page_table::KernelMode> =
+        unsafe { PageTable::from_root_register() };
+    if error_code.contains(PageFaultErrorCode::PRESENT) {
+        // FIXME: We should define the initialize mapping and the protect method here should not change the
+        // permission of the initialize mapping.
+        //
+        // Safety: The page fault address has been checked and the flags is constructed based on error code.
+        unsafe {
+            page_table
+                .protect(page_fault_vaddr as usize, flags)
+                .unwrap();
+        }
+    } else {
+        // Safety: The page fault address has been checked and the flags is constructed based on error code.
```

Comment:
Not sure why we construct the flags from the error code. But doing so does _not_ provide any safety. For safety, we should really construct correct flags, not use flags derived from the error code.